core/lib/types/src/storage/writes/compression.rs

```diff
@@ -0,0 +1,409 @@
+
+        if size >= 31 {
+            None
+        } else {
+            Some((diff, size))
+        }
+    }
+
+    fn output_size(&self) -> Option<usize> {
+        self.get_diff_and_size().map(|(_, size)| size)
+    }
+}
+
+struct CompressionByteTransform {
+    pub new_value: U256,
+}
+
+impl CompressionMode for CompressionByteTransform {
+    fn operation_id(&self) -> usize {
+        3
+    }
+
+    fn get_diff_and_size(&self) -> Option<(U256, usize)> {
+        // Ceiling division
+        let size = (self.new_value.bits() + 7) / 8;
+
+        if size >= 31 {
+            None
+        } else {
+            Some((self.new_value, size))
+        }
+    }
+
+    fn output_size(&self) -> Option<usize> {
+        self.get_diff_and_size().map(|(_, size)| size)
+    }
+}
+
+struct CompressionByteNone {
+    pub new_value: U256,
+}
+
+impl CompressionByteNone {
+    fn new(new_value: U256) -> Self {
+        Self { new_value }
+    }
+}
+
+impl CompressionMode for CompressionByteNone {
+    fn operation_id(&self) -> usize {
+        0
+    }
+
+    fn get_diff_and_size(&self) -> Option<(U256, usize)> {
+        None
+    }
+
+    fn output_size(&self) -> Option<usize> {
+        Some(32)
+    }
+
+    fn compress_value_only(&self) -> Option<Vec<u8>> {
+        let mut buffer = [0u8; 32];
+        self.new_value.to_big_endian(&mut buffer);
+
+        Some(buffer.to_vec())
+    }
+
+    fn compress_extended(&self) -> Option<Vec<u8>> {
+        let mut res = [0u8; 33];
+
+        self.new_value.to_big_endian(&mut res[1..33]);
+        Some(res.to_vec())
+    }
+}
+
+fn default_passes(prev_value: U256, new_value: U256) -> Vec<Box<dyn CompressionMode>> {
+    vec![
+        Box::new(CompressionByteAdd {
+            prev_value,
+            new_value,
+        }),
+        Box::new(CompressionByteSub {
+            prev_value,
+            new_value,
+        }),
+        Box::new(CompressionByteTransform { new_value }),
+    ]
+}
+
+/// Generates the metadata byte for a given compression strategy.
+/// The metadata byte is structured as:
+/// First 5 bits: length of the compressed value
+/// Last 3 bits: operation id corresponding to the given compression used.
+fn metadata_byte(output_size: usize, operation_id: usize) -> u8 {
+    ((output_size << 3) | operation_id) as u8
+}
+
+/// For a given previous value and new value, try each compression strategy selecting the most
+/// efficient one. Using that strategy, generate the extended compression (metadata byte and compressed value).
+/// If none are found then use the full 32 byte new value with the metadata byte being `0x00`
+pub fn compress_with_best_strategy(prev_value: U256, new_value: U256) -> Vec<u8> {
+    let compressors = default_passes(prev_value, new_value);
+
+    compressors
+        .iter()
+        .filter_map(|e| e.compress_extended())
+        .min_by_key(|bytes| bytes.len())
+        .unwrap_or_else(|| {
+            CompressionByteNone::new(new_value)
```

Comment:
we could probably add CompressionByteNone as the last entry to the default_passes. Then you don't need to treat it as a special case.

Reply:
+1 for this

Reply:
was also thinking of this - the downside would be that we'd need to unwrap as `CompressionByteNone` returns `Some`

Reply:
but it's already unwrapped

Reply:
not in this scenario, `min_by_key` returns an `Option<Vec<u8>>`. I personally like the separation, `CompressionByteNone` feels less like a compression strategy and more of a default as there's no gain savings